src/dos/dos_keyboard_layout.cpp

```diff
@@ -39,78 +39,99 @@
 
 extern void DOS_UpdateCurrentProgramName();
 
-static void notify_code_page_changed(const bool keyboard_layout_changed = false)
+static const std::string ResourceDirKeyboard = "freedos-keyboard";
+static const std::string ResourceDirCodePage = "freedos-cpi";
+
+static void notify_code_page_changed()
 {
 	// Re-create various information to match new code page
 	DOS_UpdateCurrentProgramName();
-	DOS_RefreshCountryInfo(keyboard_layout_changed);
+	DOS_RepopulateCountryInfo();
 	AUTOEXEC_NotifyNewCodePage();
 }
 
 // A common pattern in the keyboard layout file is to try opening the requested
-// file first within DOS, then from the local path, and finally from builtin
-// resources. This function performs those in order and returns the first hit.
-static FILE_unique_ptr open_layout_file(const char *name, const char *resource_dir = nullptr)
+// file first within DOS, then from the local path.
+// This function performs those in order and returns the first hit.
+static FILE_unique_ptr open_layout_file(const std::string& name)
```

Comment:
`FILE_unique_ptr` is a bit of a questionable concept; it kinda wraps raw C file pointers into C++ unique pointers, but it's still a way too low level construct.

Do you _really_ need to pass file pointers around in a _slightly_ safer way using this wrapper? How about using a `Layout` object that is guaranteed to represent an existing layout, so it follows RAII principles and always cleans up after itself when the `Layout` object is constructed.

I just think this `FILE_unique_ptr` is the wrong level abstraction here, and it's only used in the localisation module, so you could easily refactor it into an RAII `Layout` object.

So, to be clear, constructing a `Layout` object would thrown an exception if the layout file is invalid or it doesn't exist.

Reply:
I intend to do something like this. Moreover, I intend to read all the keyboard layout files during the start (we have just 4 files, about 115KB total size) and rewrite the whole loading routine from the ground, as the current one is buggy (see https://github.com/dosbox-staging/dosbox-staging/issues/1925#issuecomment-2435280083).

If the information about all the keyboard layouts (namely, which code pages they support) is available in memory, we could further improve the `KEYB` command, for example `KEYB /list 852` could list all the keyboard layouts available for code page 852, and  `KEYB /list gk` could list all the code pages compatible with the _Greek_ keyboard layout.

But I don't want to do this now, this PR is already huge. In `dos_keyboard_layout.cpp` I mostly changed what I had to. I'll close this as out of scope of this PR.

Reply:
Yeah, that's sounds fair, job for later 😄